DATALAB-338: Add Aggregation Endpoint

Bettiher opened the pull request:
Co-authored w. Maria

First PR (still work to be done) for the aggregation endpoint in Amanita, but it's running! 
http://localhost:5001/content-aggregations?agg_property=masterBrand
http://localhost:5001/content-aggregations?agg_property=masterBrand&bucketKeys=bbc_webonly&bucketKeys=bbc_two&bucketSize=3

Few general comments (NOT A PRIORITY to fix):
- we had to change the pips_index in one of the test (test_clients_elastic_querybuilder_get_similar.py): we should think to a way to avoid to refer to the index as it's likely to change again !
- the sort part in the body queries are sensitive to the order of the arguments (test_clients_elastic_querybuilder_get_content.py and test_clients_elastic_querybuilder_get_content_aggregations.py): not really robust..!

The test coverage is only 70%, we're gonna add more tests. But we would like to know if there are major changes to do before. 
For example we would like to test the bucketKeys passed as parameter, and make sure there are values of the given property. The easiest solution would be to consider the different properties as boolean parameters (property_master_brand Y/N etc.) and to set the bucketKeys in the allowed_values in the mapping. To be discussed with expert, aka @tatiana ;)

tatiana commented:
> we had to change the pips_index in one of the test (test_clients_elastic_querybuilder_get_similar.py): we should think to a way to avoid to refer to the index as it's likely to change again !
@Bettiher & @maria-panteli one way of achieving this is to use index aliasses: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html

I would certainly recommend this approach. The alias could be called pips, and it could map the latest Amanita-compatible ES index.

> For example we would like to test the bucketKeys passed as parameter, and make sure there are values of the given property. The easiest solution would be to consider the different properties as boolean parameters (property_master_brand Y/N etc.) and to set the bucketKeys in the allowed_values in the mapping. To be discussed with expert, aka @tatiana ;)

I'd say @twindebank is the Amanita export of our team, but I can gladly share thoughts on it. Could you give some examples of how would all the query strings of both appraches work?

tatiana commented on amanita/apiparams/lists.py:
```diff
@@ -38,6 +38,21 @@ def get_param_mappers_for_endpoint(endpoint, definitions):
+            'includeBadTitle': definitions.include_bad_title,
+            'languages': definitions.languages,
+            'latest': definitions.latest
+        }
```
I'm a bit worried with the duplication of mappers - which was not introduced by this PR, but it doesn't make it better either.

It could make sense to define on the top of the function:
`collection_mappers`
With the parameters which are standard to endpoints which provide a collection of items, and re-use them when defining each mappter.
An example:
```
COLLECTION_MAPPERS = {
    'limit': definitions.limit,
    'languages': definitions.languages
}
mappers = {
    'property': definitions.property,
    'bucketKeys': definitions.bucket_keys,
    'bucketSize': definitions.bucket_size,
    'exclusionList': definitions.exclusion_list,
    **COLLECTION_MAPPERS
}
```

Bettiher replied:
We grouped 2 kinds of mappers for now. "collection" with sort, limit and offset. And "filter" with includeWeather, includeBadTitle, languages and latest. To avoid the repetitions a bit and keep some sense in the groupings. Still not perfect tho.

tatiana commented on amanita/apiparams/types.py:
```diff
@@ -50,6 +50,15 @@ def __new__(cls, integer):
         return integer
 
 
+class StrictlyPositiveFloat(float):
```
It seems this property is being used for the `BucketSize` only. Shouldn't the BucketSize be a positive integer? If it is possible, we could reuse the existing `StrictlyPositiveInt`.

Bettiher replied:
Yes true. We put a float in the specs and here to stay generic in case we want to pass on proportions but the parameter name won't be accurate anymore... Will change to integer!

Bettiher replied:
Done

tatiana commented on amanita/clients/dummy.py:
```diff
@@ -38,3 +38,10 @@ def get_multiple_items(self, uris):
+
+    def get_content_aggregations(self, property):
+        logging.debug("Sending dummy get_content_aggregations response")
+        return {'results': [{'bucketKey': 'bbc_one', 'bucketItems': [{'title': 'BBC one clip'},
```
It would be good to have the dictionary in the line before the return, in multiple lines, so we can better understand it.
Eg (but with regular identation, eg, 4 spaces etc): 
```
response = {
    "results": [
        {
            "bucketKey": "bbc_one",
            "bucketItems" : [
                     {'title': 'BBC one clip'}
             ]
           }
    ]
}
```

Bettiher replied:
done

tatiana commented on amanita/clients/elastic/client.py:
```diff
@@ -45,6 +45,12 @@ def get_similar(self, mapped_params):
 
+    def get_content_aggregations(self, mapped_params):
+        query_body = get_content_aggregations.build_query_body(**mapped_params)
+        es_response = self.query(query_body, index=ES_INDEX, doc_type='clip', scroll='1m')
```
Again, this was not introduced as part of this PR, but it would be great if we could have as constants, in the top of the file:
```
DOC_TYPE = 'clip'
SCROLL = '1m'
```
And change all the references to these values to use the constants. This way, if we want to change in future, we'd be able to do it in a single place.

Bettiher replied:
done

tatiana commented on amanita/clients/elastic/process_response.py:
```diff
@@ -18,3 +18,29 @@ def map_hits_to_api_spec(es_res):
     return clips
+
+
+def map_aggregated_hits_to_api_spec(es_res):
```
We can have some simple unittests to validate the behaviour of this function

maria-panteli replied:
added tests in test_clients_elastic_process_response.py

tatiana commented on amanita/clients/elastic/query_parameter_mappers.py:
```diff
@@ -1,6 +1,6 @@
 """MockDefinitions of API parameter validators for all endpoints for Elasticsearch only."""
 
-from amanita.apiparams.types import BoolFromString, StrictlyPositiveInt, URIStr, \
+from amanita.apiparams.types import BoolFromString, StrictlyPositiveFloat, StrictlyPositiveInt, URIStr, \
```
If we are to remove the `StrictlyPositiveFloat`, we probably don't want to import it here.

Bettiher replied:
done ;)

tatiana commented on amanita/clients/elastic/querybuilder/get_content_aggregations.py:
```diff
@@ -0,0 +1,83 @@
+        raise InvalidInputParameterCombination('Cannot specify both `sort` and `random`.')
+
+    search = Search(index=ES_INDEX)
+    search = search[offset:offset + limit]  # TODO: THIS DOESNT WORK?? query builds as should but no effect
```
if possible, update this comment saying that apparently the `offset` isn't working, although `limit` is. It seems @twindebank was stuck with this bug before, it would be great if you could give it a try to make the offset work as well - sometimes all we need is fresh eyes :)

tatiana replied:
@maria-panteli is this you could have a quick look at?

tatiana commented on amanita/clients/elastic/querybuilder/get_content_aggregations.py:
```diff
@@ -0,0 +1,83 @@
+        # specify size=1000 otherwise only 10 buckets are returned by default
+        search.aggs.bucket('property', 'terms', field='genres.id', size=1000) \
+            .metric('top_hits', 'top_hits', size=bucket_size,
+                    sort=[{"updatedTime": {"order": "desc"}}])
```
Since it seems the `pid` is the reliable way of sorting clips by the publish time (as updated time could change over time, but it doesn't necessarily mean that video is newer) - it seems we should be using `pid` to sort.

jensraaby replied:
@tatiana Is that a safe assumption to make? I don't know if there's any guarantee that they will always go up. PIDs from different sources start with different letters and that wouldn't necessarily be something you want to sort by (e.g. B, W, P).

maria-panteli replied:
I changed the aggregation query to sort by pid instead of updatedTime. Shall I change this in the default sort of the /content-aggregation and the /content endpoint?

tatiana replied:
@jensraaby I'm glad you pointed this out. Have a look here:
https://github.com/bbc/connected-data/blob/master/docs/arch/017-adr-default-version-selection-logic.md#more-than-one-version-with-the-same-type

During the implementation of this ADR:
https://github.com/bbc/datalab-connectors/pull/15

I raised a concern about using the PID for sorting, and it seems this is used by the iPlayer team. It felt unsafe, but as iPlayer was doing it, it seemed to be something the PIPS team had designed for. This was part of my comment:
```
have a function to identify the newest version (is the PID really the best way? If PIPS changes the logic of how it generates PID, we'd be pottentially selecting the wrong version. Why not scheduledStart?)
```
@jensraaby What is the recommended field by PIPS to order by recency? Should we update the logic of versions sorting to use it as well? Should the iPlayer team be informed of this, so they an follow this same strategy?

tatiana commented on amanita/clients/elastic/querybuilder/get_content_aggregations.py:
```diff
@@ -0,0 +1,83 @@
+    # add aggregation
+    if property == 'genres':
+        # specify size=1000 otherwise only 10 buckets are returned by default
+        search.aggs.bucket('property', 'terms', field='genres.id', size=1000) \
```
It seems the number `1000`  should either be a constant, or we want to compute it based on the number of how many `bucketKeys` we receive.

maria-panteli replied:
after discussing with @tatiana this size parameter will be defined as a constant and updated to len(bucketKeys) if bucketKeys are provided. If bucketKeys are not provided we would need to make a call to ES to figure out the total number of bucketKeys. This ES call doesn't look like it should live together with the build_query_body code. So, for now we would be limiting to 1000 bucketKeys if bucketKeys are not provided.

tatiana commented on amanita/clients/elastic/querybuilder/get_content_aggregations.py:
```diff
@@ -0,0 +1,83 @@
+    search = Search(index=ES_INDEX)
+    search = search[offset:offset + limit]  # TODO: THIS DOESNT WORK?? query builds as should but no effect
+
+    if sort:
```
We will probably want to review this (not only here, but in the other parts of the code)| later on. I asked the mining bees and it seems they normaly prefer short clips - that said, each sort in our query, slows down our query.

tatiana commented on amanita/clients/elastic/querybuilder/get_content_aggregations.py:
```diff
@@ -0,0 +1,83 @@
+from amanita.exceptions.queryexceptions import InvalidInputParameterCombination
+
+
+def build_query_body(property, bucket_keys=None, bucket_size=5, exclusion_list=None,
```
`property` is a reserved word in python 😬  usually it is a good idea to not use it in variables. Still, given the current logic of Amanita, this is the way to do it. Could we consider changing this query string parameter from `property` to `field`?

Bettiher replied:
We changed it to agg_property
Gonna update the specs!

jensraaby commented on amanita/clients/elastic/process_response.py:
```diff
@@ -18,3 +18,29 @@ def map_hits_to_api_spec(es_res):
+        bucket = {'bucketKey': agg_item.get('key'),
+                  'bucketItems': clips}
+        buckets.append(bucket)
+    return buckets
```
The `clip` dict is identical to the one in the previous function in `process_response.py` . Any chance you could extract that functionality out? I think making the inner for loop smaller would aid readability (it might even be possible to make it into a list comprehension).

Bettiher replied:
done

tatiana commented on amanita/clients/elastic/querybuilder/get_content_aggregations.py:
```diff
@@ -0,0 +1,83 @@
+            sort.append('-defaultVersion.duration')
+        search = search.sort(*sort)
+
+    # todo: check if property is an object in a better way..
```
@maria-panteli & @Bettiher  Wouldn't it be more flexible, and require less code, if who's using Amanita sends as `bucketKey` already "genres.id" and "programmeFormats"? We can document this in the specs in Swagger, to make it clear gthe usage. This way Amanita would be flexible and we would not have any magic to handle fields which are strings or keywords x fields which are properties.

Bettiher replied:
Done + specs changed in the architecture repo

maria-panteli replied:
good idea, changed the allowed values for the agg_property to be masterBrand, genres.id, programmeFormats.id. and the query has less if-statements now!

tatiana commented on amanita/settings.py:
```diff
@@ -14,4 +14,4 @@
 DB_PASS = os.getenv('DB_PASS')
 DB_CLIENT = os.getenv('DB_CLIENT', 'elasticsearch')
-ES_INDEX = os.getenv('ES_INDEX', 'pips')
+ES_INDEX = os.getenv('ES_INDEX', 'pips_v2')
```
If we have `pips` as the ES alias, than we wouldn't need to change the settings whenever we create a new index compatible with Amanita.

jensraaby commented on tests/test_clients_elastic/test_clients_elastic_querybuilder_get_content_aggregations.py:
```diff
@@ -0,0 +1,115 @@
+from amanita.exceptions.queryexceptions import InvalidInputParameterCombination
+
+params = json.load(open("tests/test_clients_elastic/data/param_examples.json"))
+property = params['property']['validated']
```
Same property name warning here

jensraaby commented on tests/test_clients_elastic/test_clients_elastic_querybuilder_get_similar.py:
```diff
@@ -60,7 +60,7 @@ def test_query_building_all_implemented_params():
                                                                            'synopses'],
                                                                 'like': {'_id': 'http://exampleuri.com',
-                                                                         '_index': 'pips',
+                                                                         '_index': 'pips_v2',
```
_aside_: Have we got a good way to deal with this changing? It seems we will be changing the index a few times soon with the new APPW mapping that's coming

tatiana replied:
@jensraaby it seems the best way for us to handle this would be using:
https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html
As I mentioned in a comment above. Whenever we have a new index in ES which is compatible with Amanita, we could just shift the `pips` alias  to map the newest version. We would just need to decide when to do this - would this be a manual step, would we be doing this in the ingest process...?

Bettiher commented on amanita/clients/elastic/query_parameter_mappers.py:
```diff
@@ -73,3 +73,23 @@
     is_list=False,
     allowed_values=['true', 'false']
 )
+property = ParameterMapper(
```
done

Bettiher replied:
changed to agg_property

jensraaby commented:
I lost the thread about sorting by PID - I'm sure someone asked about it somewhere.
I think if we can guarantee we are only seeing PIDs that start with the same letter then sorting by that should be good enough for now. Otherwise we may need a better way to do it, perhaps using availability times.

Bettiher commented:
@tatiana @jensraaby we addressed most of the comments.

What is remaining is: 
- we will change the index to the alias when it will be set up in the ES instance 
- the offset is still not working in the queries -will have a closer look at that 

Can you let us know if you are happy with the changes otherwise? Thanks!

tatiana commented:
@Bettiher & @maria-panteli would it be possible to update Drone so it also checks PEP8 as well? I've just made this change to Bramble:
https://github.com/bbc/connected-data-bramble/blob/f947f5a61b8c3709179e842cbb3618b446bce5e2/.drone.yml#L8
:)

tatiana commented:
@maria-panteli & @Bettiher it may be worth to cover these lines with tests:
```
amanita/clients/elastic/querybuilder/get_content_aggregations.py      36      3     26      2    89%   26-28, 23->26, 62->65
```

tatiana commented on amanita/apiparams/lists.py:
```diff
@@ -6,38 +6,50 @@
+        'includeBadTitle': definitions.include_bad_title,
+        'languages': definitions.languages,
+        'latest': definitions.latest
+    }
```
👍 this really helps, @Bettiher @maria-panteli , thanks for updating it!

tatiana commented on amanita/apiparams/lists.py:
```diff
@@ -6,38 +6,50 @@
+            'random': definitions.random,
+            **collection_mappers,
+            **filter_mappers
+        }
```
❤️

tatiana commented:
@janae locally all the tests passed for me - and I also get the two warnings you mentioned, which were already in master branch. I believe those two classes were never meant to be tests on PyCharm, and were unluckily have Test in their names 😬

tatiana commented:
@Bettiher and @maria-panteli wonderful work! With this PR, both of you earned the "Amanita" badges 🥇 
I've successfully run all the tests. There are some PEP8 issues, as described before - it would be great if you could address them! I also mentioned above how to make sure our CI is checking PEP8 as well.
Last but not the least, I successfully run Amanita locally, with the commands:
```
curl 'http://localhost:5001/content-aggregations?aggProperty=genres.id&bucketKeys=100012&bucketKeys=100007&bucketSize=1'

curl 'http://localhost:5001/content-aggregations?aggProperty=masterBrand&bucketKeys=bbc_radio_bristol&bucketSize=4'
```
And I also run Snowdrop using this version of Amanita and made sure all the endpoints used by the app remain working after the current changes.

I'm assuming you'll do the PEP8 minor changes and I'm approving this PR, you deserve it! 🎉

Bettiher commented:
As explained to @janae the tests fail because of the index (if not set up to pips_v2 it will fail), will be fixed with the alias!

janae commented:
A different thought, looking at the current output I get from the aggregation endpoint, it contains all the information about each item. Shouldn't it just return a minimum set of fields that we really need? uri, bucket id, and updated time (for sorting)?

Bettiher commented:
@janae we decided to stay generic, it might be use for something else than personalisation. And sticked to the same shape for the media items as the other endpoints

janae commented:
:shipit: 
I've been using the branch a reasonable amount for dev and it behaves well
